### ISSUE TITLE
Dependency management

### ISSUE THREAD

**Issue by kafitzgerald** [NCAR/miles-credit#264]
Currently, all dependencies are specified in pyproject.toml under `dependencies`.  This includes things that are developer focused or only needed for certain use cases.  

Leveraging optional dependencies and dependency groups could be quite helpful here and hopefully thin down the list of required dependencies as well.

**Comment by djgagne** [comment]
I agree with reorganizing and removing dependencies to reduce the bulk of creating the environment. 

Some dependencies to move to optional for development purposes:
* pytest
* sphinx
* sphinx-autoapi
* myst_parser
* jupyter (has a ton of submodule dependencies)
* ruff

Optional depedencies for application extensions:
* miles-guess (only used for p-type)
* hagelslag (used in miles guess for metrics; currently not used anywhere in CREDIT)
* keras (is pretty lightweight but only used for miles-guess models)

Optional for distributed training and distributed calculation of solar but also hardware dependent
* mpi4py

Optional for certain model types:
* pytorch_geometric
* pytorch segmentation models

Some dependencies to remove entirely because they are used very little and come with additional baggage:
* torchmetrics (not listed in requirements.txt only used as a metric in diffusion)
* pysteps (only used for rank histogram; causes some installation issues on mac and requires extra installation time to build from source when installed with pip or uv)
* torchvision (only used in the current iteration of transforms to compose steps together, could be replaced with a nn.Sequential or a new workflow)

 Dependencies I'd like to get rid of but aren't sure how:
* xESMF (It requires conda to build ESMPy but is still the most convenient option for regridding GFS data. Other regridding tools are just really slow, harder to use, or not as comprehensive. The best alternative option I have found is ESMF_RegridWeightGen, which creates a weights netCDF file from two appropriately formatted grid files and supports all the common forms of regridding. Applying the regridding is really trivial once you have the weights (just a matmul), but the weights do take some time to calculate, especially if you want to do it right.)

**Comment by djgagne** [comment]
Closing this for now since we have reduced dependencies in gen2.
